agent/consul/rate/handler.go

```diff
@@ -0,0 +1,158 @@
+// package rate implements server-side RPC rate limiting.
+package rate
+
+import (
+	"context"
+	"errors"
+	"net"
+	"sync/atomic"
+
+	"github.com/hashicorp/consul/agent/consul/rate/multilimiter"
+)
+
+var (
+	// ErrRetryElsewhere indicates that the operation was not allowed because the
+	// rate limit was exhausted, but may succeed on a different server.
+	//
+	// Results in a RESOURCE_EXHAUSTED or "429 Too Many Requests" response.
+	ErrRetryElsewhere = errors.New("rate limit exceeded, try a different server")
+
+	// ErrRetryLater indicates that the operation was not allowed because the rate
+	// limit was exhausted, and trying a different server won't help (e.g. because
+	// the operation can only be performed on the leader).
+	//
+	// Results in an UNAVAILABLE or "503 Service Unavailable" response.
+	ErrRetryLater = errors.New("rate limit exceeded, try again later")
+)
+
+// Mode determines the action that will be taken when a rate limit has been
+// exhausted (e.g. log and allow, or reject).
+type Mode int
+
+const (
+	// ModePermissive causes the handler to log the rate-limited operation but
+	// still allow it to proceed.
+	ModePermissive Mode = iota
+
+	// ModeEnforce causes the handler to reject the rate-limted operation.
+	ModeEnforce
+)
+
+// OperationType is the type of operation the client is attempting to perform.
+type OperationType int
+
+const (
+	// OperationTypeRead represents a read operation.
+	OperationTypeRead OperationType = iota
+
+	// OperationTypeWrite represents a write operation.
+	OperationTypeWrite
+)
+
+// Operation the client is attempting to perform.
+type Operation struct {
+	// Name of the RPC endpoint (e.g. "Foo.Bar" for net/rpc and "/foo.service/Bar" for gRPC).
+	Name string
+
+	// SourceAddr is the client's (or forwarding server's) IP address.
+	SourceAddr net.Addr
+
+	// Type of operation to be performed (e.g. read or write).
+	Type OperationType
+}
+
+// Handler enforces rate limits for incoming RPCs.
+type Handler struct {
```

Comment:
We will need to have a logger and be able to emit metrics in the handler.

Reply:
Yep! I'm aware that we're planning to do some buffering and wasn't sure whether the handler would do this, or we'd encapsulate it in a wrapper around the logger (which would be my preference).

Anyway, we can tackle that in a future PR 🙇🏻

Reply:
Yep! I'm aware that we're planning to do some buffering and wasn't sure whether the handler would do this, or we'd encapsulate it in a wrapper around the logger (which would be my preference).

Anyway, we can tackle that in a future PR 🙇🏻